feat(trampoline): add a trampoline module (@endo/trampoline) (#2263)
endojs/endo: packages/trampoline/README.md

# supertramp

> Sync and async trampolines, for your pleasure

## Wat

From [Wikipedia](https://en.wikipedia.org/wiki/Trampoline_(computing)):

> > A trampoline is a loop that iteratively invokes [thunk](https://en.wikipedia.org/wiki/Thunk_(functional_programming))-returning functions ([continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style)).

Author: To be clear, you're _not_ saying, "don't call it a trampoline", yes?  There is [precedent](https://www.npmjs.com/search?q=trampoline) in the JS ecosystem for this meaning. 

I would prefer we name it something silly, of course.

Open to suggestions.  I like `rainbow-trampoline`, because it supports functions of any color 🌈.

> > A trampoline is a loop that iteratively invokes [thunk](https://en.wikipedia.org/wiki/Thunk_(functional_programming))-returning functions ([continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style)).

Reviewer: Function coloring is also a loaded term. Sync/async coloring is currently the main kind in JS, but coloring has been used in the context of a couple other proposals lately.

> > A trampoline is a loop that iteratively invokes [thunk](https://en.wikipedia.org/wiki/Thunk_(functional_programming))-returning functions ([continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style)).

Author: I can, without rancor, use the more-widely-understood meaning of words within their context (in this case, the community of JavaScript developers) while disregarding those less-widely-understood meanings. Naming is hard and we do the best we can 😄 

But if you (or anyone else) has a suggestion for naming this package, please offer it!

> > A trampoline is a loop that iteratively invokes [thunk](https://en.wikipedia.org/wiki/Thunk_(functional_programming))-returning functions ([continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style)).

Reviewer 2: `optional-a-sync`

> > A trampoline is a loop that iteratively invokes [thunk](https://en.wikipedia.org/wiki/Thunk_(functional_programming))-returning functions ([continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style)).

Reviewer 2: @mhofman @boneskull need more razors for this yak shaving?

I tend to use GPT for seeding name ideas. gpt-4o is really bad at it tho, had to go back to gpt-4
Here are some not awful seeds which we could use as inspiration for coming up with something reasonable.
```
1. AsyncBridge
2. SyncAsyncHarmonizer
3. JumpSync
4. YieldBounce
5. GenTrampolinator
6. AsyncOrNot
8. CodeChameleon
12. SyncAsyncLeap
16. AsyncFlip
17. YieldFlex
18. BounceBetween
19. SyncAsyncFusion
```

I still stand by `optional-a-sync` but these could also work from above:
`async-or-not` `sync-async-chameleon` `sync-async-bridge`

> > A trampoline is a loop that iteratively invokes [thunk](https://en.wikipedia.org/wiki/Thunk_(functional_programming))-returning functions ([continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style)).

Reviewer 3: I would like to name this `trampoline`.

> > A trampoline is a loop that iteratively invokes [thunk](https://en.wikipedia.org/wiki/Thunk_(functional_programming))-returning functions ([continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style)).

Author: I'd name it "The Right Honourable Boris von Meija" if it were totally up to me, but I can live with `@endo/trampoline`.

 */
function* loadRecursive(thunk, filename) {
  let specifiers = yield thunk(filename);
  
  // pretend there's some de-duping, caching, 
  // scrubbing, etc. happening here!

  for (const specifier of specifiers) {
    specifiers = [...specifiers, ...(yield* loadRecursive(thunk, specifier))];
  }
  return specifiers;
}

// results are an array of all imports found in some.js' dependency tree
const asyncResult = await trampoline(loadRecursive, readAsync, './some.js');
const syncResult = syncTrampoline(loadRecursive, readSync, './some.js');

asyncResult === syncResult; // true
```

In the above example, **@endo/trampoline** allows us to re-use the recursive operations in `loadRecursive()` for _both_ sync and async execution. An implementation _without_ **@endo/trampoline** would need to duplicate the operations into two (2) discrete recursive functions—a synchronous-colored function and an asynchronous-colored function. Over time, this situation commonly leads to diverging implementations. If that _doesn't_ sound like a big deal for _whatever you're trying to do here_, then you probably don't need **@endo/trampoline**.

## What is this?

> The pattern exposed by this library—known as [trampolining][]—helps manage control flow in a way that avoids deep recursion and potential stack overflows. It effectively "converts" recursive calls into a loop. This is especially helpful in a language like JavaScript [because reasons][proper-tail-calls].

Reviewer 2: This line remains slightly confusing - first time reading it I didn't grok that it's describing what the library is NOT.
```suggestion
This library is not just an implementation of a trampoline that helps manage control flow in a way that avoids deep recursion and potential stack overflows (by turning recursive calls into a loop). 
```

Author: for posterity: even if it is not an explicit goal, we get mitigation of the lack of proper TCO for free. right?

**@endo/trampoline** provides the trampolining pattern, but in such a way that a consumer can execute _either_ synchronous _or_ asynchronous operations _paired with operations common to both_.

> In other words, **@endo/trampoline** can help _reduce code duplication_ when recursive operations must be executed _in both sync and async_ contexts.

Reviewer 2: ```suggestion
In other words, **@endo/trampoline** can help _reduce code duplication_ when complex logic, potentially even with recursive operations must be executed _in both sync and async_ contexts with patterns like dependency injection used to provide matching sync or async utilities to the main logic.
```
